Support login with role_id and secret_id

mhodovaniuk opened the pull request:
Add role secret_id support to login.

Fixes #302

rossabaker commented:
Thanks!  I just approved the new build, but I think we're still going to get binary compatibility exceptions.  This library has a stable API, so we need to be careful about breaking changes.

There are a couple things we can do here: 
* add overloads that take the new parameter, so the original signatures are still there
* create a "builder" pattern, so we can do something like `login.withRoleSecretId(...).execute`.  (The uses of [MeterBuilder in otel4s](https://github.com/typelevel/otel4s/blob/c25617ca2b8409f99e6b5ab348728fafce3c4226/core/src/main/scala/org/typelevel/otel4s/otel4s.scala#L82-L86), and its [implementation](https://github.com/typelevel/otel4s/blob/main/java/src/main/scala/org/typelevel/otel4s/java/impl.scala#L43-L59), are an example of what I'm talking about.)  It's a bit more code, but it's more futureproof as new parameters get added.

rossabaker commented:
Oh, the build passed because of the MiMa exceptions.  Yeah, we're going to need to find a way to make this pass without those.  Does that otel4s example make sense how the pattern might be applied here?  I can give more detail if it doesn't.

mhodovaniuk commented:
Thanks for the feedback!
Implemented `loginAndKeep` and `loginAndKeepSecretLeased` with builder pattern. Also, implemented `login` with builder pattern to make API consistent.

rossabaker commented:
Sorry, got distracted, but tagging myself so I can look at again tomorrow when I'm fresh.

mhodovaniuk commented:
@rossabaker a gentle reminder on this.

rossabaker commented on core/src/main/scala/com/banno/vault/Vault.scala:
```diff
@@ -333,4 +322,59 @@ object Vault {
+    def apply(roleId: String)(implicit F: Concurrent[F]): F[VaultToken]
+  }
+
+  final case class VaultLoginOperationImpl[F[_]](client: Client[F], vaultUri: Uri, roleSecretId: Option[String] = None) extends VaultLoginOperation[F] {
```
When this is a case class, we can't add new parameters later in a bincompat fashion,  so this isn't quite the builder pattern.  Making it an abstract class with a private constructor should work.  You'll need to implement your own `copy` then.  I also don't think this needs to be public.

mhodovaniuk replied:
@rossabaker thanks you for getting back to this
May I kindly ask you to provide some examples of similar implementations with "abstract class with a private constructor"? 
I tried to get back to the examples that you provided before to get some insights, but it looks like the project was refactored and now it uses [case class](https://github.com/typelevel/otel4s/blob/main/java/metrics/src/main/scala/org/typelevel/otel4s/java/metrics/MeterBuilderImpl.scala).

rossabaker commented on core/src/main/scala/com/banno/vault/Vault.scala:
```diff
@@ -333,4 +322,59 @@ object Vault {
+    def apply(roleId: String, tokenLeaseExtension: FiniteDuration)(implicit A: Async[F]): Stream[F, String]
+  }
+
+  final case class VaultLoginAndKeepOperationImpl[F[_]](client: Client[F], vaultUri: Uri, roleSecretId: Option[String] = None) extends VaultLoginAndKeepOperation[F] {
```
Same comments as VaultLoginOperationImpl.

rossabaker reviewed (changes requested):
Sorry it took me so long to get back to this.

We still have those MiMa exceptions that we need to get rid of.  I wonder if we could make the old signatures private, and delegate to your new implementations.

This looks mostly correct, but a couple comments on the builder pattern.

Koroeskohr commented:
Hello! I'm looking into this feature, I'd be willing to finish this MR if we can agree this is the implementation that should be targeted :)

mhodovaniuk commented:
Good luck @Koroeskohr !
As an alternative, I can recommend https://github.com/jopenlibs/vault-java-driver. It is a zero-dependency Java client. For example, my code is on the ZIO stack, so this was a huge plus. Additionally, the maintainer is very supportive; I was missing a small feature, and they merged my PR very quickly

morgen-peschke commented:
This is supported as of v9.4.0, sorry for the delay.

morgen-peschke commented:
This is supported as of v9.4.0, sorry for the delay.
